### PR TITLE
[fix][flaky-test]NamespaceServiceTest.flaky/testModularLoadManagerRemoveBundleAndLoad

### LINKED ISSUE #17386: Flaky-test: NamespaceServiceTest. testModularLoadManagerRemoveBundleAndLoad
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Example failure

https://github.com/apache/pulsar/runs/8126481494?check_suite_focus=true

### Exception stacktrace

<details>
<summary>Full exception stacktrace</summary>
<code><pre>
  Error:  Tests run: 39, Failures: 1, Errors: 0, Skipped: 33, Time elapsed: 54.468 s <<< FAILURE! - in org.apache.pulsar.broker.namespace.NamespaceServiceTest
  Error:  testModularLoadManagerRemoveBundleAndLoad(org.apache.pulsar.broker.namespace.NamespaceServiceTest)  Time elapsed: 3.879 s  <<< FAILURE!
  java.lang.AssertionError: expected [true] but found [false]
  	at org.testng.Assert.fail(Assert.java:99)
  	at org.testng.Assert.failNotEquals(Assert.java:1037)
  	at org.testng.Assert.assertTrue(Assert.java:45)
  	at org.testng.Assert.assertTrue(Assert.java:55)
  	at org.apache.pulsar.broker.namespace.NamespaceServiceTest.testModularLoadManagerRemoveBundleAndLoad(NamespaceServiceTest.java:766)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
  	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
  	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
  	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
  	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
  	at java.base/java.lang.Thread.run(Thread.java:833)
</pre></code>
</details>

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by poorbarcode**
Fixes #17386

### Motivation

In this test, we expect it works like this: 

1. persistent broker data to ZK
2. receive notice: broker data changes, and refresh bundle data in memory 
3. persistent bundle data which already updated in memory to ZK
4. verify bundle data persist correctly
5. delete namespace
6. verify bundle data deleted

But the flow works like this:
1. persistent broker data to ZK
2. <strong> (High light) race condition </strong>
2-x. receive notice: broker data changes, and refresh bundle data in memory 
2-x. persistent bundle data to ZK ( If the data in memory is not updated, it will not be written to ZK )
3. ......

### Modifications

- guarantee `step-3` is always executed after `step-2`
- In this test, the original codes mock `ModularLoadManagerWrapper` to make `isCentralized` always return `true`. But this mock is completely unnecessary because that's `ModularLoadManagerWrapper.isCentralized` was exactly always returned `true`, so remove the mock. 

### Documentation

- [ ] `doc-required` 
  
- [x] `doc-not-needed` 
  
- [ ] `doc` 

- [ ] `doc-complete`

### CODE REVIEW

**Comment by poorbarcode** [comment]
This PR should merge to `branch-2.10` and `branch-2.11` and `master`

**Comment by codelipenghui** [comment]
@heesung-sn Please help review this PR, thanks.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/namespace/NamespaceServiceTest.java [hunk 1]
```diff
@@ -792,12 +794,46 @@ public void testModularLoadManagerRemoveBundleAndLoad() throws Exception {
...
-        loadManager.writeResourceQuotasToZooKeeper();
+        waitResourceDataUpdateToZK(loadManager,
+                loadData -> !loadData.getBundleData().containsKey(bundleName));
 
         getResult = pulsar.getLocalMetadataStore().get(path).get();
         assertFalse(getResult.isPresent());
+    }
 
+    private void waitResourceDataUpdateToZK(
+            LoadManager loadManager, Predicate<LoadData> utilChecker) throws Exception {
```
Comment (gaoran10) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/namespace/NamespaceServiceTest.java:805]: It seems the param `utilChecker` is useless.

**Review by gaoran10 — APPROVED** [review]
Great work! LGTM

**Review by Technoboy- — APPROVED** [review]
Better to add some comments for line 809~814

**Comment by poorbarcode** [comment]
Hi @Technoboy- 

> Better to add some comments for line 809~814

Already add code-comment, thanks

**Comment by heesung-sn** [comment]
LGTM

**Comment by heesung-sn** [comment]
// Wait until "ModularLoadManager" completes processing the ZK notification.
        ModularLoadManagerWrapper modularLoadManagerWrapper = (ModularLoadManagerWrapper) loadManager;
        ModularLoadManagerImpl modularLoadManager = (ModularLoadManagerImpl) modularLoadManagerWrapper.getLoadManager();
        ScheduledExecutorService scheduler = Whitebox.getInternalState(modularLoadManager, "scheduler");
        
        
        We don't have the `scheduler` anymore in ModularLoadManagerImpl. I think this test is failing now.

**Pull request merged**